fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1065,6 +1065,74 @@ void assertResolutionStateMutationsSizeConsistent(const std::vector<ResolveTrans
 	}
 }
 
+// Return true if a single-key mutation is associated with a valid tenant id or a system key
+bool validTenantAccess(MutationRef m, std::unordered_map<int64_t, TenantName> const& tenantMap) {
+	if (isSystemKey(m.param1))
+		return true;
+
+	if (isSingleKeyMutation((MutationRef::Type)m.type)) {
+		auto tenantId = TenantAPI::extractTenantIdFromMutation(m);
+		if (tenantMap.count(tenantId)) {
+			return true;
+		} else {
+			return false;
+		}
+	}
+	return true;
+}
+
+inline bool tenantMapChanging(MutationRef mutation) {
+	const KeyRangeRef tenantMapRange = TenantMetadata::tenantMap().subspace;
+	if (mutation.type == MutationRef::SetValue && mutation.param1.startsWith(tenantMapRange.begin)) {
+		return true;
+	} else if (mutation.type == MutationRef::ClearRange &&
+	           tenantMapRange.intersects(KeyRangeRef(mutation.param1, mutation.param2))) {
+		return true;
+	}
+	return false;
+}
+
+// Return success and properly split clear range mutations if all tenant check pass. Otherwise, return corresponding
+// error
+Error validateAndProcessTenantAccess(VectorRef<MutationRef> mutations, ProxyCommitData* const pProxyCommitData) {
+	bool changeTenant = false;
+	bool writeNormalKey = false;
+	for (auto& mutation : mutations) {
+		if (tenantMapChanging(mutation)) {
+			changeTenant = true;
+		} else if (mutation.type == MutationRef::ClearRange) {
+			auto beginTenantId = TenantAPI::extractTenantIdFromKeyRef(mutation.param1);
+			auto endTenantId = TenantAPI::extractTenantIdFromKeyRef(mutation.param2);
+			CODE_PROBE(beginTenantId != endTenantId, "Clear Range raw access or cross multiple tenants");
+			// TODO: splitting clear range
+		} else if (!isSystemKey(mutation.param1)) {
+			if (!validTenantAccess(mutation, pProxyCommitData->tenantMap)) {
+				return illegal_tenant_access();
+			}
+			writeNormalKey = true;
+		}
+
+		if (writeNormalKey && changeTenant) {
```

Comment:
is it possible to change the tenant map and write a mutation in the same CommitTransactionRequest?

Reply:
No. We explicitly disallow that.

Reply:
Is this the location where we disallow this? If not then why do we need this check?

Reply:
Yes, CP needs to check change the tenant map and write a mutation to normal key space. If that's true, it treat the transaction as an illegal transaction.